llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -0,0 +1,230 @@
+    // inside the module's entry point.
+    // YKFIXME: Investigate languages that don't have/use main as the first
+    // entry point.
+    Function *Main = M.getFunction("main");
+    if (Main == nullptr) {
+      Context.emitError("Unable to add shadow stack: could not find \"main\" function!");
+      return false;
+    }
+    Instruction *First = Main->getEntryBlock().getFirstNonPHI();
+    IRBuilder<> Builder(First);
+
+    // Now create some memory on the heap for the shadow stack.
+    FunctionCallee MF = M.getOrInsertFunction("malloc", Int8PtrTy, PointerSizedIntTy);
+    CallInst *Malloc = Builder.CreateCall(MF,
+        {ConstantInt::get(PointerSizedIntTy, SHADOW_STACK_SIZE)}, "");
+    Builder.CreateStore(Malloc, GShadowStackPtr);
+
+    Value *SSPtr;
+    for (Function &F : M) {
+      if (F.empty()) // skip declarations.
+        continue;
+
+      if (&F != Main) {
+        // At the top of each function in the module, load the heap pointer
+        // from the global shadow stack variable.
+        Builder.SetInsertPoint(F.getEntryBlock().getFirstNonPHI());
+        SSPtr = Builder.CreateLoad(Int8PtrTy, GShadowStackPtr);
+      } else {
+        SSPtr = cast<Value>(Malloc);
+      }
+
+      size_t Offset = 0;
+      // Remember which allocas were replaced, so we can remove them later in
+      // one swoop. Removing them here messes up the loop.
+      std::vector<Instruction *> RemoveAllocas;
+      for (BasicBlock &BB: F) {
+        for (Instruction &I: BB) {
+          if (isa<AllocaInst>(I)) {
+            // Replace allocas with pointers into the shadow stack.
+            AllocaInst &AI = cast<AllocaInst>(I);
+            if (isYkMTNewAlloca(&AI)) {
+              // The variable created by `yk_mt_new` will never be traced, so
+              // there's no need to store it on the shadow stack.
+              continue;
+            }
+            if (isa<StructType>(AI.getAllocatedType())) {
+              StructType *ST = cast<StructType>(AI.getAllocatedType());
+              // Some yk specific variables that will never be traced and thus
+              // can live happily on the normal stack.
+              if (ST->getName() == "YkCtrlPointVars" || ST->getName() == "struct.YkLocation") {
+                continue;
+              }
+            }
+            Builder.SetInsertPoint(&I);
+            auto AllocaSizeInBits = AI.getAllocationSizeInBits(DL);
+            if (!AllocaSizeInBits) {
+              // YKFIXME: Deal with functions where the stack size isn't know at
+              // compile time, e.g. when `alloca` is used.
+              Context.emitError("Unable to add shadow stack: function has dynamically sized stack!");
+              return false;
+            }
+            // Calculate this `AllocaInst`s size, aligning its pointer if
+            // necessary, and create a replacement pointer into the shadow
+            // stack.
+            size_t AllocaSize = *AllocaSizeInBits / 8;
```

Comment:
`/8` should probably be the word size and not `8`.

Reply:
Fixed in [7c4306c](https://github.com/ykjit/ykllvm/pull/53/commits/7c4306c3fc7894c6d03e933c429842029b2a4f91)